Design Picker: Show Blank Canvas on every category, remove "No Category", and add the "Show All" filter

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

Make some adjustments to the category picker now that theme categories have been updated.

* Show "featured" (e.g. Blank Canvas) themes at the top of every category
* Remove the "No Category" filter. If a theme doesn't have a category, and it's not "featured" then it won't appear in the design picker
* Adds a "Show All" filter to the category sidebar

<img width="1556" alt="Screenshot 2021-11-12 at 8 00 36 PM" src="https://user-images.githubusercontent.com/1500769/141424530-4d6403c6-d559-4098-bf49-554d19636bf9.png">



#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Test on `calypso.localhost` so that the category sidebar feature flag is enabled
* Go through the hero flow and not the changes have been made to the design picker

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #56567

p-jackson committed:
Show Blank Canvas in every category

p-jackson committed:
Remove "No Category" category

p-jackson committed:
Add "Show All" filter to category picker

arthur791004 commented on client/signup/steps/design-picker/index.jsx:
```diff
@@ -86,9 +86,10 @@ class DesignPickerStep extends Component {
-					{ name: this.props.translate( 'No Category' ), slug: 'CLIENT_ONLY-no-category' },
-				],
+				categories: taxonomies?.theme_subject ?? [],
+				// Blank Canvas uses the theme_picks taxonomy with a "featured" term in order to
```
Is Blank Canvas the only one to use `theme_picks` with a "featured" term?

p-jackson replied:
Yes at the moment it is. This is the logic the gallery at `/themes/:siteSlug` uses to show Blank Canvas at the top, so I thought it'd be good to replicate that logic rather than special case the `blank-canvas` slug.

arthur791004 commented on packages/design-picker/src/utils/index.ts:
```diff
@@ -13,6 +13,13 @@ export function gatherCategories( designs: Design[] ): Category[] {
 }
 
+// Returns the category slug to use for the special "Show All" category
+export function getShowAllSlug(): string {
```
I think we can just define a variable as followed. Or is there any reason we should use the function 🤔
```js
const SHOW_ALL_SLUG = 'DESIGN_PICKER_SHOW_ALL_SLUG';
```

p-jackson replied:
Perhaps a constant makes more sense. Not sure why I went with a function 🤷

arthur791004 commented on packages/design-picker/src/components/index.tsx:
```diff
@@ -220,9 +223,18 @@ const DesignPicker: React.FC< DesignPickerProps > = ( {
+	const { __ } = useI18n();
+
 	const categories = gatherCategories( designs );
+	if ( showAllFilter && designs.length ) {
```
Why do we need to check the `designs.length` here?

IMO, changing `gatherCategories` to `useGatherCategories( designs, showAllFilter )` might make this component more simple. But it's okay to go with the current way.

p-jackson replied:
> Why do we need to check the `designs.length` here?

During loading the All category would show even when there were no designs loaded yet. Checking `designs.length` just ensures that the category will only appear if it won't be an empty category.

p-jackson replied:
I like that `gatherCategories()` is a simple (i.e. non-hook) helper function. But you're right that the logic in the component's render function is getting a bit dense. There could be a `useCategorization()` sort of hook that I could extract out of it. Something that dealt with category selection, chooses the default selection, that sort of thing.

arthur791004 commented on packages/design-picker/src/components/index.tsx:
```diff
@@ -241,6 +253,8 @@ const DesignPicker: React.FC< DesignPickerProps > = ( {
 		? designs
 		: filterDesignsByCategory( designs, selectedCategory );
 
+	filteredDesigns.sort( sortDesigns );
```
IMO, sorting in client/signup/steps/design-picker/index.jsx:L86 might be better as we don't need this on gutenboarding.

p-jackson replied:
I added it here because it's part of the `showFirst` logic. Do you think it'd be better to have the design step be responsible for adding Blank Canvas to every category?
It's possible, but we'd need to artificially add every category to the Blank Canvas theme, and the design step doesn't really have any concept of categories at the moment (i.e. `gatherCategories` is in the DP component)

arthur791004 replied:
Oh I don't notice that you've added the `showFirst` to the type definition of `Design`. But I still think we should avoid sorting at rendering. It's okay for now as the size of `filteredDesigns` is small. However, the `sort` function might cause the side effect easily because it mutate the original array. If we need to use it here, I prefer to clone the value first.
```suggestion
	const sortedDesign = filteredDesigns.slice().sort( sortDesigns );
```

p-jackson replied:
I hadn't noticed how I was mutating the prop that was passed in if `! showCategoryFilter`! That feels like poor form and it should be cloned. Perhaps I'll add a `useMemo()` too since I don't need to sort for _every_ render.

p-jackson replied:
[Turns out `.slice` is the most consistently fast method of cloning on different browsers](https://perf.link/#eyJpZCI6Im41dXZhZ3Z6aXdxIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGRhdGEgPSBbXVxuZm9yKGxldCBpID0gMDsgaSA8IDEwMDAwOyBpKyspIHsgZGF0YS5wdXNoKE9iamVjdC5jcmVhdGUoe30pKSB9IiwidGVzdHMiOlt7Im5hbWUiOiJzbGljZSIsImNvZGUiOiJjb25zdCByZXN1bHQgPSBkYXRhLnNsaWNlKDApIiwicnVucyI6WzEzMzc1LDE0MTI1LDE2MDAwLDEzMTI1LDExNTAwLDEzODc1LDE0MTI1LDE0MDAwLDE0MjUwLDEzNTAwLDk3NTAsMTIyNTAsMTgyNTAsMTQ2MjUsMTMzNzUsMTM2MjUsMTQwMDAsMTMzNzUsMTM2MjUsNDUwMCwxMzM3NSwxMzM3NSwxMzUwMCwxMzYyNSwyMTg3NSwxNDAwMCwyMTI1MCwxODEyNSwxMzc1MCwyMDEyNSwxMzM3NSwxMDUwMCwxMTYyNSwyNzM3NSwxMjI1MCwxMzM3NSwxMzg3NSwxMzc1MCwxOTAwMCwxMzM3NSwxMzUwMCwxMzM3NSwxMzc1MCwxMzc1MCwxMzc1MCwxMzg3NSwxNDAwMCwxMzYyNSwxMTI1MCwxNTM3NSwxNjc1MCwxMzc1MCwxMDg3NSwxMzI1MCwxMjUwMCwxNDM3NSwxMDM3NSwxMzM3NSwyMjM3NSwxODI1MCwxMjYyNSwxNDAwMCwxMzAwMCwxNDAwMCw5MDAwLDEzNTAwLDIxMDAwLDEzNTAwLDE3MTI1LDEwMjUwLDEzMzc1LDEyMjUwLDEzNzUwLDEyMjUwLDE2NjI1LDExMjUwLDE5MDAwLDEzMzc1LDEzNjI1LDExNjI1LDEzMTI1LDEzNzUwLDEzNzUwLDEzNTAwLDEzMzc1LDEzMzc1LDEzNjI1LDEzNTAwLDExODc1LDUzNzUsMTM2MjUsMTMzNzUsMjA4NzUsMTM1MDAsMTM1MDAsMTc2MjUsMTM2MjUsMTMzNzUsMTE4NzUsMTAzNzVdLCJvcHMiOjE0MDEwfSx7Im5hbWUiOiJzcHJlYWQiLCJjb2RlIjoiY29uc3QgcmVzdWx0ID0gWy4uLmRhdGFdIiwicnVucyI6WzUwMCwxMzc1LDEwMDAsMTc1MCwxMzc1LDEyNSwxNTAwLDIzNzUsMTg3NSwyMzc1LDEzNzUsMTYyNSw2MjUsMTUwMCwxNTAwLDEzNzUsMTEyNSwyMDAwLDEyNSwxODc1LDE4NzUsNzUwLDM3NSw2MjUsMzEyNSwxODc1LDEzNzUsMTM3NSwyNTAsMTM3NSwxNTAwLDEzNzUsMTM3NSwyNTAwLDEzNzUsMTM3NSwxMzc1LDE4NzUsMzc1LDM3NSwyNTAwLDYyNSw4NzUsNTAwLDEzNzUsMjYyNSwxMzc1LDUwMCwyMzc1LDEzNzUsMTM3NSwxNTAwLDEzNzUsMTM3NSwxNTAwLDE2MjUsMTM3NSwzNzUsNzUwLDEzNzUsNTAwLDEzNzUsMTM3NSwxMzc1LDEzNzUsMTAwMCwxMjUsMTI1LDEwMDAsMTM3NSwxMzc1LDE2MjUsMTI1LDEzNzUsMTM3NSwxNTAwLDEzNzUsMTAwMCwxMjUwLDExMjUsMTI1LDEzNzUsMTI1LDE4NzUsMTI1LDEwMDAsMTEyNSwxMjUsMTM3NSwyMDAwLDEwMDAsMTI1LDI1MCwxMTI1LDEzNzUsMTM3NSwxMzc1LDEyNTAsMTM3NSwxNTAwXSwib3BzIjoxMjIzfSx7Im5hbWUiOiIuZnJvbSIsImNvZGUiOiJjb25zdCByZXN1bHQgPSBBcnJheS5mcm9tKGRhdGEpIiwicnVucyI6WzUwMCw4NzUsNTAwLDc1MCw4NzUsODc1LDEyNTAsODc1LDEyNTAsNzUwLDc1MCwzNzUsODc1LDM3NSwxNTAwLDEwMDAsMTI1LDc1MCwyMzc1LDEwMDAsMTAwMCwyMDAwLDEwMDAsNTAwLDg3NSwyMDAwLDg3NSwxMjUwLDE1MDAsNzUwLDI1MCwxMjUwLDg3NSwyMDAwLDIwMDAsNjI1LDYyNSw3NTAsNTAwLDc1MCwxMDAwLDc1MCw3NTAsNzUwLDEwMDAsNzUwLDg3NSw3NTAsODc1LDc1MCwxMjUwLDg3NSwxMDAwLDg3NSw3NTAsODc1LDc1MCwyMDAwLDc1MCw4NzUsMTAwMCw3NTAsNzUwLDg3NSw4NzUsNzUwLDI1MCw3NTAsNjI1LDEyNTAsODc1LDg3NSw4NzUsODc1LDc1MCw4NzUsMjUwLDEwMDAsODc1LDUwMCwxMDAwLDg3NSw3NTAsNjI1LDg3NSw4NzUsMTEyNSwzNzUsODc1LDg3NSw3NTAsMjUwLDI1MCw1MDAsNzUwLDM3NSw3NTAsMzc1LDM3NSw4NzVdLCJvcHMiOjg2M31dLCJ1cGRhdGVkIjoiMjAyMS0xMS0xNVQxOTo0MToyMS42NDRaIn0%3D). TIL

arthur791004 commented:
I found that if the user previews the theme and then goes back to the design picker, the selected category would be reset and the user experience is not great. If we can keep the category, it would be better.

simison commented:
@ollierozdarz @ianstewart Would it make sense to do same as we do with page template picker, and have "blank" option above categories? I wonder if we would see popularity of blank options go down a bit and if that's a good thing or not?

<img width="854" alt="image" src="https://user-images.githubusercontent.com/87168/141482287-70b5043d-a1ca-48e2-82de-d81101c58d4c.png">

Button label could be more elaborate, "start with blank canvas" or something like that.

p-jackson commented:
> I found that if the user previews the theme and then goes back to the design picker, the selected category would be reset and the user experience is not great. If we can keep the category, it would be better.

Good catch, yes I think this should be addressed

p-jackson committed:
Refactored the category handling code into a custom hook

p-jackson committed:
Pre-select the Blog category if user signals intention to "write"

p-jackson committed:
Refactor design picker into a functional component

This will allow use to use a hook from @automattic/design-picker within
the design picker signup step.

p-jackson committed:
Move Design Picker category selection state up into signup step

arthur791004 reviewed (commented):
Everything looks good except this

p-jackson committed:
Memoize design sorting

p-jackson commented:
@arthur791004 thanks for the reviews so far. I've done a fairly large refactor switching the design picker step component into a functional component. Apologies for the big diff, hopefully it makes more sense when you read individual commits.

In order to have the category selection persist when navigating between "step sections" I needed to push the selection state further up the component tree into the step component, and to ensure it shared the logic with `@automattic/design-picker` I needed `DesignPickerStep` to work with hooks. Which is why I migrated it to a functional component.

The scroll management and tracks events side effects appear to still be working after switching from lifecycle methods to `useEffect`, but probably worth double checking for me.

ollierozdarz commented:
> @ollierozdarz @autumnfjeld @ianstewart Would it make sense to do same as we do with page template picker, and have "blank" option above categories? I wonder if we would see popularity of blank options go down a bit and if that's a good thing or not?

@simison, I think it depends if we want to devalue the option or not. Maybe if we can get the numbers on the E2E success (i.e. let's say those who launch) of users who start with a blank page vs those who don't, that's the answer in making it more subtle or leaving it as is.

ollierozdarz commented:
@p-jackson may be unrelated to this specific ticket, but can we please put the Blogs category at the top, and active by default, since that's +61% of our user base?

arthur791004 commented on client/signup/steps/design-picker/index.jsx:
```diff
@@ -250,76 +254,65 @@ class DesignPickerStep extends Component {
-		}
+	const isMobile = useViewportMatch( 'small', '<' );
+
+	const { flowName, stepName, isReskinned } = props;
```
I personally prefer to do object destructuring at the beginning of the component.

p-jackson replied:
You're right that's more usual. Just a hangover from refactoring the `render()` method I guess.

arthur791004 reviewed (approved):
LGTM. Well done for refactoring to the functional component 👍

p-jackson committed:
Move prop destructuring

p-jackson commented:
> @p-jackson may be unrelated to this specific ticket, but can we please put the Blogs category at the top, and active by default, since that's +61% of our user base?

@ollierozdarz I've pinged you on this conversation p1636700899343200-slack-C029SB8JT8S about which category should be the default. This PR doesn't deploy the category picker to users so I'll merge this and follow up with a small PR with the conclusion of that conversation. I'm happy to always make "Blog" the default.

ollierozdarz commented:
@p-jackson thanks! Replied in the Slack thread.
